apps/test/unit/lib/ui/accounts/ChangeUserTypeControllerTest.js

```diff
@@ -0,0 +1,149 @@
+      setupWithInitialUserType(INITIAL_USER_TYPE);
+    });
+
+    it('button is initially disabled', () => {
+      expect(button.prop('disabled')).to.be.true;
+    });
+
+    it('changing dropdown enables/disables button', () => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+      expect(button.prop('disabled')).to.be.false;
+
+      dropdown.val(INITIAL_USER_TYPE);
+      dropdown.change();
+      expect(button.prop('disabled')).to.be.true;
+    });
+
+    it('clicking the enabled button shows a modal dialog', () => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+
+      expect(ReactDOM.render).not.to.have.been.called;
+      button.click();
+      expect(ReactDOM.render).to.have.been.calledOnce;
+    });
+
+    it('can hide the modal dialog', () => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+      button.click();
+      expect(ReactDOM.render).to.have.been.calledOnce;
+
+      expect(ReactDOM.unmountComponentAtNode).not.to.have.been.called;
+      controller.hideChangeUserTypeModal();
+      expect(ReactDOM.unmountComponentAtNode).to.have.been.calledOnce;
+    });
+  });
+
+  describe('when initialUserType = "teacher"', () => {
+    const INITIAL_USER_TYPE = 'teacher';
+    const OTHER_USER_TYPE = 'student';
+
+    beforeEach(() => {
+      setupWithInitialUserType(INITIAL_USER_TYPE);
+    });
+
+    it('button is initially disabled', () => {
+      expect(button.prop('disabled')).to.be.true;
+    });
+
+    it('changing dropdown enables/disables button', () => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+      expect(button.prop('disabled')).to.be.false;
+
+      dropdown.val(INITIAL_USER_TYPE);
+      dropdown.change();
+      expect(button.prop('disabled')).to.be.true;
+    });
+
+    it('clicking button submits form, reloads page on success', (done) => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+      button.click();
+      expect(form.submit).to.have.been.calledOnce;
+
+      expect(dropdown.prop('disabled')).to.be.true;
+      expect(button.prop('disabled')).to.be.true;
+      expect(status.text()).to.equal(i18n.saving());
+
+      utils.reload.callsFake(() => done());
+      form.trigger('ajax:success');
+    });
+
+    it('re-enables controls on failure and displays error', (done) => {
+      dropdown.val(OTHER_USER_TYPE);
+      dropdown.change();
+      button.click();
+      expect(form.submit).to.have.been.calledOnce;
+
+      expect(dropdown.prop('disabled')).to.be.true;
+      expect(button.prop('disabled')).to.be.true;
+      expect(status.text()).to.equal(i18n.saving());
+
+      form.trigger('ajax:error', [{}]);
+      // Hidden promise needs time to resolve...
+      setTimeout(() => {
+        expect(dropdown.prop('disabled')).to.be.false;
+        expect(button.prop('disabled')).to.be.false;
+        expect(status.text()).to.equal(i18n.changeUserTypeModal_unexpectedError());
+        done();
+      }, 2);
```

Comment:
Safe to use a timeout in this way?

Reply:
Safe?  Yes, totally fine in a test.  Reliable in the long term? Maybe not, but I wasn't happy with the change I'd need to make to the class under test to get a handle on the actual promise.  Do you see a good way to do it?

Reply:
Can we use [Chai as Promised](http://www.chaijs.com/plugins/chai-as-promised/), like here:?
https://github.com/code-dot-org/code-dot-org/blob/d3fe64057b71494a610d1c282f8c3d5279e4834d/apps/test/unit/templates/SchoolAutocompleteDropdownTest.js#L150-L154

Reply:
My problem is that I don't currently have a handle on the promise that we'd need to pass into `expect()`.

Reply:
OIC. I missed that part in your earlier comment. [This](https://github.com/code-dot-org/code-dot-org/pull/22403/files#diff-68b7abab9c23a3fc3ca757c9f0a41177R66) promise? 

Would it be crazy to assign that to a class variable? Then the test could inspect and return something like `controller.instance().submitPromise`?

Reply:
[How does this look?](https://github.com/code-dot-org/code-dot-org/pull/22403/commits/6bcef5c8747790054ffcc87a1ff37be850225631)

Reply:
That looks great! 🎉  It's so much cleaner with async/await than Chai as Promised.